File: app/src/main/java/com/drdisagree/iconify/ui/widgets/SliderWidget.java
package com.drdisagree.iconify.ui.widgets;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.drdisagree.iconify.R;
import com.google.android.material.slider.Slider;

import java.text.DecimalFormat;
import java.util.Objects;

@SuppressLint("CustomViewStyleable")
public class SliderWidget extends RelativeLayout {

    private LinearLayout container;
    private TextView titleTextView;
    private TextView summaryTextView;
    private Slider materialSlider;
    private ImageView resetIcon;
    private String valueFormat;
    private int defaultValue;
    private float outputScale = 1f;
    private boolean isDecimalFormat = false;
    private String decimalFormat = "#.#";
    private OnLongClickListener resetClickListener;
    private Slider.OnSliderTouchListener onSliderTouchListener;

    public SliderWidget(Context context) {
        super(context);
        init(context, null);
    }

    public SliderWidget(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public SliderWidget(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        inflate(context, R.layout.view_widget_slider, this);

        initializeId();

        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.CustomWidgetView);
        valueFormat = a.getString(R.styleable.CustomWidgetView_valueFormat);
        defaultValue = a.getInt(R.styleable.CustomWidgetView_sliderDefaultValue, Integer.MAX_VALUE);
        setTitle(a.getString(R.styleable.CustomWidgetView_titleText));
        setSliderValueFrom(a.getInt(R.styleable.CustomWidgetView_sliderValueFrom, 0));
        setSliderValueTo(a.getInt(R.styleable.CustomWidgetView_sliderValueTo, 100));
        setSliderStepSize(a.getInt(R.styleable.CustomWidgetView_sliderStepSize, 1));
        setSliderValue(a.getInt(
                R.styleable.CustomWidgetView_sliderValue,
                a.getInt(R.styleable.CustomWidgetView_sliderDefaultValue, 50)
        ));
        isDecimalFormat = a.getBoolean(R.styleable.CustomWidgetView_isDecimalFormat, false);
        decimalFormat = a.getString(R.styleable.CustomWidgetView_decimalFormat);
        outputScale = a.getFloat(R.styleable.CustomWidgetView_outputScale, 1f);
        a.recycle();

        if (valueFormat == null) {
            valueFormat = "";
        }

        if (decimalFormat == null) {
            decimalFormat = "#.#";
        }

        setSelectedText();
        handleResetVisibility();
    }

    public void setTitle(int titleResId) {
        titleTextView.setText(titleResId);
    }

    public void setTitle(String title) {
        titleTextView.setText(title);
    }

    public void setSelectedText() {
        summaryTextView.setText(
                (valueFormat.isBlank() || valueFormat.isEmpty() ?
                        getContext().getString(
                                R.string.opt_selected1,
                                String.valueOf(
                                        !isDecimalFormat ?
                                                (int) (materialSlider.getValue() / outputScale) :
                                                new DecimalFormat(decimalFormat)
                                                        .format(materialSlider.getValue() / outputScale)
                                )
                        ) :
                        getContext().getString(
                                R.string.opt_selected2,
                                !isDecimalFormat ?
                                        String.valueOf((int) materialSlider.getValue()) :
                                        new DecimalFormat(decimalFormat)
                                                .format(materialSlider.getValue() / outputScale),
                                valueFormat
                        )
                )
        );
    }

    public void setSliderStepSize(int value) {
        materialSlider.setStepSize(value);
    }

    public void setSliderValue(int value) {
        materialSlider.setValue(value);
        setSelectedText();
        handleResetVisibility();
    }

    public int getSliderValue() {
        return (int) materialSlider.getValue();
    }

    public void setSliderValueFrom(int value) {
        materialSlider.setValueFrom(value);
    }

    public void setSliderValueTo(int value) {
        materialSlider.setValueTo(value);
    }

    public void setIsDecimalFormat(boolean isDecimalFormat) {
        this.isDecimalFormat = isDecimalFormat;
        setSelectedText();
    }

    public void setDecimalFormat(String decimalFormat) {
        this.decimalFormat = Objects.requireNonNullElse(decimalFormat, "#.#");
        setSelectedText();
    }

    public void setOutputScale(float scale) {
        this.outputScale = scale;
        setSelectedText();
    }

    public void setOnSliderTouchListener(Slider.OnSliderTouchListener listener) {
        onSliderTouchListener = listener;

        materialSlider.addOnSliderTouchListener(new Slider.OnSliderTouchListener() {
            @Override
            public void onStartTrackingTouch(@NonNull Slider slider) {
                notifyOnSliderTouchStarted(slider);
            }

            @Override
            public void onStopTrackingTouch(@NonNull Slider slider) {
                setSelectedText();
                handleResetVisibility();
                notifyOnSliderTouchStopped(slider);
            }
        });
    }

    public void setOnSliderChangeListener(Slider.OnChangeListener listener) {
        materialSlider.addOnChangeListener(listener);
    }

    public void setResetClickListener(OnLongClickListener listener) {
        resetClickListener = listener;

        resetIcon.setOnLongClickListener(v -> {
            if (defaultValue == Integer.MAX_VALUE) {
                return false;
            }

            setSliderValue(defaultValue);
            handleResetVisibility();
            notifyOnResetClicked(v);

            return true;
        });
    }

    private void notifyOnSliderTouchStarted(@NonNull Slider slider) {
        if (onSliderTouchListener != null) {
            onSliderTouchListener.onStartTrackingTouch(slider);
        }
    }

    private void notifyOnSliderTouchStopped(@NonNull Slider slider) {
        if (onSliderTouchListener != null) {
            onSliderTouchListener.onStopTrackingTouch(slider);
        }
    }

    private void notifyOnResetClicked(View v) {
        if (resetClickListener != null) {
            resetClickListener.onLongClick(v);
        }
    }

    private void handleResetVisibility() {
        if (defaultValue != Integer.MAX_VALUE && materialSlider.getValue() != defaultValue) {
            resetIcon.setVisibility(VISIBLE);
        } else {
            resetIcon.setVisibility(GONE);
        }
    }

    @Override
    public void setEnabled(boolean enabled) {
        super.setEnabled(enabled);

        container.setEnabled(enabled);
        titleTextView.setEnabled(enabled);
        summaryTextView.setEnabled(enabled);
        resetIcon.setEnabled(enabled);
        materialSlider.setEnabled(enabled);
    }

    // to avoid listener bug, we need to re-generate unique id for each view
    private void initializeId() {
        container = findViewById(R.id.container);
        titleTextView = findViewById(R.id.title);
        summaryTextView = findViewById(R.id.summary);
        materialSlider = findViewById(R.id.slider_widget);
        resetIcon = findViewById(R.id.reset);

        container.setId(View.generateViewId());
        titleTextView.setId(View.generateViewId());
        summaryTextView.setId(View.generateViewId());
        materialSlider.setId(View.generateViewId());
        resetIcon.setId(View.generateViewId());
    }
}

File: app/src/main/java/com/drdisagree/iconify/ui/widgets/SliderWidget.kt
package com.drdisagree.iconify.ui.widgets

import android.content.Context
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import com.drdisagree.iconify.R
import com.drdisagree.iconify.utils.HapticUtils.weakVibrate
import com.google.android.material.button.MaterialButton
import com.google.android.material.slider.Slider
import kotlinx.parcelize.Parcelize
import kotlinx.parcelize.RawValue
import java.text.DecimalFormat
import java.util.Objects

class SliderWidget : RelativeLayout {

    private lateinit var container: LinearLayout
    private lateinit var titleTextView: TextView
    private lateinit var summaryTextView: TextView
    private lateinit var materialSlider: Slider
    private lateinit var resetButton: MaterialButton
    private var valueFormat: String? = ""
    private var defaultValue = 0
    private var outputScale = 1f
    private var isDecimalFormat = false
    private var decimalFormat: String? = "#.#"
    private var resetClickListener: OnLongClickListener? = null
    private var onSliderTouchListener: Slider.OnSliderTouchListener? = null

    constructor(context: Context) : super(context) {
        init(context, null)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(context, attrs)
    }

    private fun init(context: Context, attrs: AttributeSet?) {
        inflate(context, R.layout.view_widget_slider, this)

        initializeId()

        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.SliderWidget)

        valueFormat = typedArray.getString(R.styleable.SliderWidget_valueFormat)
        defaultValue = typedArray.getInt(R.styleable.SliderWidget_sliderDefaultValue, Int.MAX_VALUE)
        setTitle(typedArray.getString(R.styleable.SliderWidget_titleText))
        setSliderValueFrom(typedArray.getInt(R.styleable.SliderWidget_sliderValueFrom, 0))
        setSliderValueTo(typedArray.getInt(R.styleable.SliderWidget_sliderValueTo, 100))
        setSliderStepSize(typedArray.getInt(R.styleable.SliderWidget_sliderStepSize, 1))
        sliderValue = typedArray.getInt(
            R.styleable.SliderWidget_sliderValue,
            typedArray.getInt(R.styleable.SliderWidget_sliderDefaultValue, 50)
        )
        isDecimalFormat = typedArray.getBoolean(R.styleable.SliderWidget_isDecimalFormat, false)
        decimalFormat = typedArray.getString(R.styleable.SliderWidget_decimalFormat)
        outputScale = typedArray.getFloat(R.styleable.SliderWidget_outputScale, 1f)

        typedArray.recycle()

        if (valueFormat == null) {
            valueFormat = ""
        }

        if (decimalFormat == null) {
            decimalFormat = "#.#"
        }

        setSelectedText()
        handleResetButton()
        setOnSliderTouchListener(null)
        setResetClickListener(null)
    }

    fun setTitle(titleResId: Int) {
        titleTextView.setText(titleResId)
    }

    fun setTitle(title: String?) {
        titleTextView.text = title
    }

    fun setSelectedText() {
        summaryTextView.text =
            if (valueFormat!!.isBlank() || valueFormat!!.isEmpty()) context.getString(
                R.string.opt_selected1, (
                        if (!isDecimalFormat) (materialSlider.value / outputScale).toInt() else DecimalFormat(
                            decimalFormat
                        ).format((materialSlider.value / outputScale).toDouble())
                        ).toString()
            ) else context.getString(
                R.string.opt_selected2,
                if (!isDecimalFormat) materialSlider.value.toInt().toString() else DecimalFormat(
                    decimalFormat
                ).format((materialSlider.value / outputScale).toDouble()),
                valueFormat
            )
    }

    fun setSliderStepSize(value: Int) {
        materialSlider.stepSize = value.toFloat()
    }

    var sliderValue: Int
        get() = materialSlider.value.toInt()
        set(value) {
            materialSlider.value = value.toFloat()
            setSelectedText()
            handleResetButton()
        }

    fun setSliderValueFrom(value: Int) {
        materialSlider.valueFrom = value.toFloat()
    }

    fun setSliderValueTo(value: Int) {
        materialSlider.valueTo = value.toFloat()
    }

    fun setIsDecimalFormat(isDecimalFormat: Boolean) {
        this.isDecimalFormat = isDecimalFormat
        setSelectedText()
    }

    fun setDecimalFormat(decimalFormat: String) {
        this.decimalFormat = Objects.requireNonNullElse(decimalFormat, "#.#")
        setSelectedText()
    }

    fun setOutputScale(scale: Float) {
        outputScale = scale
        setSelectedText()
    }

    fun setOnSliderTouchListener(listener: Slider.OnSliderTouchListener?) {
        onSliderTouchListener = listener

        materialSlider.addOnSliderTouchListener(object : Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {
                notifyOnSliderTouchStarted(slider)
            }

            override fun onStopTrackingTouch(slider: Slider) {
                setSelectedText()
                handleResetButton()
                notifyOnSliderTouchStopped(slider)
            }
        })

        materialSlider.setLabelFormatter {
            if (valueFormat!!.isBlank() || valueFormat!!.isEmpty()) (if (!isDecimalFormat) (materialSlider.value / outputScale).toInt() else DecimalFormat(
                decimalFormat
            )
                .format((materialSlider.value / outputScale).toDouble())).toString() + valueFormat else (if (!isDecimalFormat) materialSlider.value.toInt()
                .toString() else DecimalFormat(decimalFormat)
                .format((materialSlider.value / outputScale).toDouble())) + valueFormat
        }
    }

    fun setOnSliderChangeListener(listener: Slider.OnChangeListener) {
        materialSlider.addOnChangeListener(listener)
    }

    fun setResetClickListener(listener: OnLongClickListener?) {
        resetClickListener = listener

        resetButton.setOnClickListener { v: View ->
            if (defaultValue == Int.MAX_VALUE) {
                return@setOnClickListener
            }

            sliderValue = defaultValue

            handleResetButton()
            notifyOnResetClicked(v)
        }
    }

    fun resetSlider() {
        resetButton.performLongClick()
    }

    private fun notifyOnSliderTouchStarted(slider: Slider) {
        onSliderTouchListener?.onStartTrackingTouch(slider)
    }

    private fun notifyOnSliderTouchStopped(slider: Slider) {
        onSliderTouchListener?.onStopTrackingTouch(slider)
    }

    private fun notifyOnResetClicked(v: View) {
        v.weakVibrate()
        resetClickListener?.onLongClick(v)
    }

    private fun handleResetButton() {
        if (defaultValue != Int.MAX_VALUE) {
            resetButton.visibility = VISIBLE
            resetButton.isEnabled = materialSlider.value != defaultValue.toFloat()
        } else {
            resetButton.visibility = GONE
        }
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)

        container.isEnabled = enabled
        titleTextView.isEnabled = enabled
        summaryTextView.isEnabled = enabled
        resetButton.isEnabled = enabled
        materialSlider.isEnabled = enabled
    }

    // to avoid listener bug, we need to re-generate unique id for each view
    private fun initializeId() {
        container = findViewById(R.id.container)
        titleTextView = findViewById(R.id.title)
        summaryTextView = findViewById(R.id.summary)
        materialSlider = findViewById(R.id.slider_widget)
        resetButton = findViewById(R.id.reset_button)
        container.id = generateViewId()
        titleTextView.id = generateViewId()
        summaryTextView.id = generateViewId()
        materialSlider.id = generateViewId()
        resetButton.id = generateViewId()
    }

    override fun onSaveInstanceState(): Parcelable {
        val superState = super.onSaveInstanceState()
        return SavedState(superState, materialSlider.value)
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        if (state !is SavedState) {
            super.onRestoreInstanceState(state)
            return
        }
        super.onRestoreInstanceState(state.superState)
        materialSlider.value = state.sliderValue
        setSelectedText()
        handleResetButton()
    }

    @Parcelize
    class SavedState(
        private val parentState: @RawValue Parcelable?,
        val sliderValue: Float
    ) : BaseSavedState(parentState)
}
